### PR TITLE
make link handler API public

### LINKED ISSUE #2168: [LinkHandler] make the LinkHandler API public
Define the public API for the LinkHandler.

See discussion and other links relating to this topic at <0>.

<0> https://github.com/adobe/aem-core-wcm-components/discussions/1861

### PR DESCRIPTION
**Pull request by jckautzmann**
- draft API

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #2168
| Patch: Bug Fix?          |
| Minor: New Feature?      |
| Major: Breaking Change?  |
| Tests Added + Pass?      | No
| Documentation Provided   | No
| Any Dependency Changes?  |
| License                  | Apache License, Version 2.0

### CODE REVIEW

File: bundles/core/pom.xml [hunk 1]
```diff
@@ -141,6 +141,7 @@
                                     <goal>check</goal>
                                 </goals>
                                 <configuration>
+                                    <skip>true</skip>
```
Comment (bpauli) [inline bundles/core/pom.xml:144]: Can you please remove this config so it's not getting accidentally merged? You can also skip the check with the parameter `-Djacoco.skip=true` during development.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/LinkBuilder.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    Resource getResource();
+
+    /**
+     * Sets the resource holding the link properties.
+     *
+     * @param resource The resource holding the link properties.
+     * @return the link builder holding the link configuration.
+     */
+    LinkBuilder setResource(Resource resource);
```
Comment (bpauli) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/LinkBuilder.java:40]: Page, Resource and references which are used to create the Link should be passed as arguments to different constructors as they are exclusive. It makes no sense for example to build the Link with setResource and setTarget on one LinkBuilder instance.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/LinkBuilder.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+/**
+ * Interface used to hold the configuration of a link.
+ */
+public interface LinkBuilder {
+
+    /**
+     * Returns the resource holding the link properties.
+     */
+    Resource getResource();
```
Comment (bpauli) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/LinkBuilder.java:32]: The Resource should be a implementation detail of the LinkBuilder

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/LinkBuilder.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.Map;
+
+import org.apache.sling.api.resource.Resource;
+
+import com.adobe.cq.wcm.core.components.models.Page;
+
+/**
+ * Interface used to hold the configuration of a link.
+ */
+public interface LinkBuilder {
```
Comment (bpauli) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/LinkBuilder.java:27]: Public `build` method is missing which returns the Link is missing.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/LinkConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.Map;
+
+import org.apache.sling.api.resource.Resource;
+
+import com.adobe.cq.wcm.core.components.models.Page;
+
+/**
+ * Interface used to hold the configuration of a link.
+ */
+public interface LinkConfiguration {
```
Comment (kwin) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/LinkConfiguration.java:27]: It is completely unclear from method signatures and javadoc which methods are returning non-null under which circumstances. Also shouldn't those method rather return Optional?

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/DefaultLinkConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.services.link;
+
+import java.util.Map;
+
+import org.apache.sling.api.resource.Resource;
+
+import com.adobe.cq.wcm.core.components.models.Page;
+
+public class DefaultLinkConfiguration implements LinkConfiguration {
```
Comment (kwin) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/DefaultLinkConfiguration.java:24]: wouldn't it make more sense to have three different implementations with some inheritance?
PageLinkConfiguration -> ResourceLinkConfiguration -> UrlLinkConfiguration

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/LinkHandler.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.Optional;
+
+import org.jetbrains.annotations.NotNull;
+
+import com.adobe.cq.wcm.core.components.commons.link.Link;
+
+/**
+ * Interface for resolving links.
+ */
+public interface LinkHandler {
```
Comment (kwin) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/services/link/LinkHandler.java:27]: LinkFactory maybe?

**Review by kwin — CHANGES_REQUESTED** [review]

**Pull request closed without merging**